File: sources/Example_metadirective.2.c
/*
* @@name: metadirective.2c
* @@type: C
* @@compilable: yes
* @@linkable: no
* @@expect: success
* @@version: omp_5.0
*/

#define N 100
#include <stdio.h>
#include <omp.h>

void work_on_chunk(int idev, int i);

int main()                    //Driver
{
   int i,idev;

   for (idev=0; idev<omp_get_num_devices(); idev++)
   {
      #pragma omp target device(idev)
      #pragma omp metadirective \
                  when( implementation={vendor(nvidia)}, device={arch("kepler")}: \
                        teams num_teams(512) thread_limit(32) )                   \
                  when( implementation={vendor(amd)},    device={arch("fiji"  )}: \
                        teams num_teams(512) thread_limit(64) )                   \
                  default(                                                        \
                        teams)
      #pragma omp distribute parallel for
      for (i=0; i<N; i++) work_on_chunk(idev,i);
   }
   return 0;
}

File: sources/Example_metadirective.2.f90
! @@name: metadirective.2f90
! @@type: F-free
! @@compilable: yes
! @@linkable: no
! @@expect: success
! @@version: omp_5.0

program main                    !!Driver
  use omp_lib
  implicit none
  integer, parameter :: N=1000
  external           :: work_on_chunk
  integer            :: i,idev

  do idev=0,omp_get_num_devices()-1

    !$omp target device(idev)
    !$omp begin metadirective &
    !$omp&  when( implementation={vendor(nvidia)}, device={arch("kepler")}: &
    !$omp&        teams num_teams(512) thread_limit(32) )                   &
    !$omp&  when( implementation={vendor(amd)},    device={arch("fiji"  )}: &
    !$omp&        teams num_teams(512) thread_limit(64) )                   &
    !$omp&  default(                                                        &
    !$omp&        teams)
    !$omp distribute parallel for
    do i=1,N  
       call work_on_chunk(idev,i)
    end do
    !$omp end metadirective
    !$omp end target

  end do

end program

